### ISSUE TITLE
Rule proposal: `no-empty-file`

### ISSUE THREAD

**Issue by manovotny** [sindresorhus/eslint-plugin-unicorn#1477]
There seems to be all kinds of "empty" rules for [block statements](https://eslint.org/docs/rules/no-empty), [destructuring](https://eslint.org/docs/rules/no-empty-pattern), [functions](https://eslint.org/docs/rules/no-empty-function), lines between things, etc., but interestingly nothing for entirely empty files, unless I am completely overlooking something (please correct me if I am wrong and I will humbly 🤦‍♂️).

I am more than happy to submit a PR if approved.

## Description

> Disallow empty files.

Disallow files containing absolutely nothing, only whitespace, or only comments.

### No space or all whitespace

```js

```

### Only Return characters / new lines (or combination thereof) 

```js
\n
```

```js
\r
```

```js
\r\n
```

### Only comments

```js
// comment
```

```js
/* comment */
```

## Pass

```js
const value = 'string';
```

**Comment by fisker** [comment]
These should also consider "empty"

Only directive

```js
'use strict';
```

Only `EmptyStatement`s

```js
;;
```

**Comment by sindresorhus** [comment]
Yes, "empty" meaning no actual code.

**Comment by manovotny** [comment]
Good callouts @fisker! I am on board with that.

**Comment by fisker** [comment]
I saw you already have an implementation for this rule, have you figured why it can't work for json files?

**Comment by manovotny** [comment]
@fisker haha, ya caught me! 😊 Nice job on the sleuthing into my own repos! 🕵️‍♂️ I figured I'd take a stab at suggesting / moving some rules here for broader use and more resiliency. Unicorn, as a project, is structurally setup to be _*far*_ more "battle tested" with the way it generates eslint selectors, unit tests, integration tests, etc. Plus, you all think of things (like you did above) that I would never think of. I love the great minds and the community here and figured it was better to share / contribute. I hope that's ok. I would obviously deprecate my own rule and point it here, if accepted and merged.

In short, no. I haven't figured out why. One wouldn't typically be linting JSON files with Eslint, unless you're using [eslint-plugin-json](https://github.com/azeemba/eslint-plugin-json), which [it's own readme FAQ](https://github.com/azeemba/eslint-plugin-json#faqs) calls out it's own oddity with sections like [Why doesn't this plugin use `eslint` itself or just `JSON.parse`?](https://github.com/azeemba/eslint-plugin-json#why-doesnt-this-plugin-use-eslint-itself-or-just-jsonparse) and [Is `eslint` really the best tool to lint my JSON?](https://github.com/azeemba/eslint-plugin-json#is-eslint-really-the-best-tool-to-lint-my-json).

Using a `console.log` on the `Program` node, eslint seems to be unable to distinguish the difference between a "full" or empty JSON file, which is why I excluded `.json` file extensions in the first place. 

Logged output is below.

```
❯ yarn lint

yarn run v1.22.11

$ eslint . --ext .js --ext .json

empty-js-file.js Node {
  type: 'Program',
  start: 0,
  end: 0,
  loc: SourceLocation {
    start: Position { line: 1, column: 0 },
    end: Position { line: 1, column: 0 },
    filename: undefined,
    identifierName: undefined
  },
  range: [ 0, 0 ],
  leadingComments: undefined,
  trailingComments: undefined,
  innerComments: undefined,
  extra: undefined,
  comments: [],
  tokens: [],
  sourceType: 'module',
  body: [],
  parent: null
}

empty-json-file.json Node {
  type: 'Program',
  start: 0,
  end: 0,
  loc: SourceLocation {
    start: Position { line: 1, column: 0 },
    end: Position { line: 1, column: 0 },
    filename: undefined,
    identifierName: undefined
  },
  range: [ 0, 0 ],
  leadingComments: undefined,
  trailingComments: undefined,
  innerComments: undefined,
  extra: undefined,
  comments: [],
  tokens: [],
  sourceType: 'module',
  body: [],
  parent: null
}

not-empty-json-file.json Node {
  type: 'Program',
  start: 0,
  end: 0,
  loc: SourceLocation {
    start: Position { line: 1, column: 0 },
    end: Position { line: 1, column: 0 },
    filename: undefined,
    identifierName: undefined
  },
  range: [ 0, 0 ],
  leadingComments: undefined,
  trailingComments: undefined,
  innerComments: undefined,
  extra: undefined,
  comments: [],
  tokens: [],
  sourceType: 'module',
  body: [],
  parent: null
}

package.json Node {
  type: 'Program',
  start: 0,
  end: 0,
  loc: SourceLocation {
    start: Position { line: 1, column: 0 },
    end: Position { line: 1, column: 0 },
    filename: undefined,
    identifierName: undefined
  },
  range: [ 0, 0 ],
  leadingComments: undefined,
  trailingComments: undefined,
  innerComments: undefined,
  extra: undefined,
  comments: [],
  tokens: [],
  sourceType: 'module',
  body: [],
  parent: null
}

empty-js-file.js
  1:1  error  Disallow empty files  get-off-my-lawn/no-empty-files

empty-json-file.json
  1:2  error  Disallow empty files  get-off-my-lawn/no-empty-files

not-empty-json-file.json
  1:2  error  Disallow empty files  get-off-my-lawn/no-empty-files

package.json
  1:2  error  Disallow empty files  get-off-my-lawn/no-empty-files
```

If it were to be accepted into unicorn, I could exclude the `.json` check or we could keep it if you wanted to remain compatible with eslint-plugin-json. Or you might have a better idea that I have not thought of yet. 

I haven't dug into the eslint-plugin-json codebase or opened an issue there either. Perhaps there's a deeper root cause to uncover.

**Comment by manovotny** [comment]
@fisker I've dug into this a bit more... 

[eslint-plugin-json](https://github.com/azeemba/eslint-plugin-json) has issues because it's using a non-standard preprocessor (it's using [vscode-json-languageservice](https://www.npmjs.com/package/vscode-json-languageservice)) and currently returns an empty string, which makes it difficult for other plugins to work alongside of it (see https://github.com/azeemba/eslint-plugin-json/issues/38 for more information).

There appear to be more compatible ESLint JSON plugins, namely [eslint-plugin-json-es](https://github.com/zeitport/eslint-plugin-json-es) and [eslint-plugin-jsonc](https://github.com/ota-meshi/eslint-plugin-jsonc), which both the standard ESLint parser [espree](https://github.com/eslint/espree) to create compliant AST. I've tested both plugins and neither one create the same issues as eslint-plugin-json.

Based on all of the above, my recommendation is to:

1. Not have a `.json` extension check within the rule.
2. Recommend more compatible ESLint JSON plugins if people have issues, which could also be called out on the rule docs, if desired.

**Comment by fisker** [comment]
Thanks for the information.

> Not have a .json extension check within the rule.

Agree. Users can also ignore files with [`overrides`](https://eslint.org/docs/user-guide/configuring/configuration-files#configuration-based-on-glob-patterns)

**Comment by fisker** [comment]
This is now accepted.

**Comment by fisker** [comment]
I'm not sure how ESLint work for Vue SFC, please make sure to add test for a Vue SFC file without `<script>` block.

**Comment by manovotny** [comment]
@fisker awesome! I'll get started on an implementation. Good note on the Vue SFC too. I'll add it to the list.

**Comment by manovotny** [comment]
@fisker one last question... Name? `no-empty-file` or `no-empty-files`?

**Comment by fisker** [comment]
ESLint use `no-empty-pattern` and `no-empty-function`, so `no-empty-file`?

**Comment by manovotny** [comment]
@fisker I think I agree. Everything else seems to be singular in unicorn too. Things aren't named `arrays`, `ternaries`, etc. Singular it is.
